src/Simulation/Simulators/ToffoliSimulator/Processor.cs

```diff
@@ -0,0 +1,518 @@
+        /// Throws an exception if the rotation is invalid.
+        /// </summary>
+        /// <param name="axis">The rotation axis</param>
+        /// <param name="numerator">The numerator of the rotation angle</param>
+        /// <param name="power">The power of two for the denominator of the
+        /// rotation angle</param>
+        /// <returns>A pair of flags. The first indicates whether the rotation is
+        /// equivalent to X; the second whether or not the rotation is exactly equal
+        /// to the identity (no phase), and thus may be controlled.</returns>
+        private (bool isX, bool safe) CheckRotation(Pauli axis, long numerator, long power)
+        {
+            // Rotations around X of odd multiples of pi/2 are X (up to global phase).
+            // Rotations around any axis of multiples of pi are I (up to global phase).
+            // Rotations of any angle around I are I.
+            // Rotations around any axis of even multiples of pi are safe to control.
+            // The angle here is pi*numerator/(2^power).
+
+            // The sign of the numerator doesn't matter.
+            // Reduce the numerator and power in case there are factors of 2 in the numerator.
+            numerator = Abs(numerator);
+            while (power > 0 && numerator % 2 == 0)
+            {
+                numerator /= 2;
+                power--;
+            }
+
+            bool equivalentToX = false;
+            bool safeToControl = false;
+
+            // If the power < 0, we're guaranteed to be a multiple of 2*pi, so exactly the identity.
+            if (power < 0)
+            {
+                equivalentToX = false;
+                safeToControl = true;
+            }
+            // If power > 1, then we're an odd multiple of pi/4, so illegal
+            else if (power > 1)
+            {
+                throw new InvalidOperationException($"The Toffoli simulator can only perform rotations of multiples of pi/2.");
+            }
+            // If power == 1, then we're an odd multiple of pi/2, so X if the axis is X, 
+            // I if the axis is I, and illegal otherwise.
+            else if (power == 1)
+            {
+                switch (axis)
+                {
+                    case Pauli.PauliI:
+                        equivalentToX = false;
+                        safeToControl = false;
+                        break;
+                    case Pauli.PauliX:
+                        equivalentToX = true;
+                        safeToControl = false;
+                        break;
+                    case Pauli.PauliY:
+                    case Pauli.PauliZ:
+                        throw new InvalidOperationException($"The Toffoli simulator can only perform non-identity rotations around the X axis.");
+                }
+            }
+            // If power == 0, then we're a multiple of pi, possible an even multiple.
+            else
+            {
```

Comment:
> { [](start = 11, length = 2)

assert(power == 0)
Code does tend to change, and these asserts come very handy.